tidy3d/updater.py

```diff
@@ -1,49 +1,119 @@
+        while self.version != CurrentVersion:
+            update_fn = UPDATE_MAP.get(self.version)
+            if update_fn is None:
+                raise SetupError(f"version {self.version} not found in update map.")
+            self.sim_dict = update_fn(self.sim_dict)
+        return self.sim_dict
+
+
+"""Update conversion functions."""
+
+# versions will be dynamically mapped in this table when the update functions are initialized.
+UPDATE_MAP = {}
+
 
+def updates_to_version(version_from_string, version_to_string):
+    """Decorates a sim_dict update function to change the version."""
 
-class Updater:
-    """Updater class. Every time a minor release is made, a function updating the previous minor
-    release to the new one must be added. The assumption is that patch versions do not need to be
-    updated.
+    # make sure the version strings are legit
+    from_version = Version.from_string(version_from_string)
+    _ = Version.from_string(version_to_string)
 
-    Example
-    -------
-    >>> old_sim_dict = {"version": "1.2.1"}
-    >>> new_sim_dict = Updater().update(old_sim_dict)
-    """
+    def decorator(update_fn):
+        """The actual decorator that gets returned by `updates_to_version('x.y.z')`"""
 
-    def __getitem__(self, key):
-        """Call appropriate upgrade function based on version string."""
-        major, minor, _ = key.split(".")
-        func_name = f"update_{major}_{minor}"
-        return getattr(self, func_name)
+        @functools.wraps(update_fn)
+        def new_update_function(sim_dict: dict) -> dict:
+            """Update function that automatically adds version string."""
 
-    @staticmethod
-    def update_1_3(sim_dict: Dict):
-        """Updates version 1.3 to 1.4."""
+            sim_dict_updated = update_fn(sim_dict)
+            sim_dict_updated["version"] = version_to_string
+            return sim_dict_updated
 
-        sim_dict["boundary_spec"] = {"x": {}, "y": {}, "z": {}}
-        for dim, pml_layer in zip(["x", "y", "z"], sim_dict["pml_layers"]):
-            sim_dict["boundary_spec"][dim]["plus"] = pml_layer
-            sim_dict["boundary_spec"][dim]["minus"] = pml_layer
-        sim_dict.pop("pml_layers")
-        sim_dict["version"] = "1.4.0"
-        return sim_dict
+        UPDATE_MAP[from_version] = new_update_function
 
-    def update(self, sim_dict: Dict):
-        """Update an old Simulation dict to the current Tidy3D version."""
+        return new_update_function
 
-        new_dict = sim_dict.copy()
-        major, minor, _ = new_dict["version"].split(".")
-        current_major, current_minor, _ = __version__.split(".")
+    return decorator
 
-        while major != current_major or minor != current_minor:
-            new_dict = self[new_dict["version"]](new_dict)
-            major, minor, _ = new_dict["version"].split(".")
 
-        new_dict["version"] = __version__
+@updates_to_version(version_from_string="1.3.0", version_to_string="1.4.0")
```

Comment:
1.3.x to 1.4.x?

Reply:
Actually I think 1.3.x to 1.4.0 makes sense and will work?

Reply:
patches are pretty much ignored, so it will work yea.